dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/RouteController.java

```diff
@@ -0,0 +1,96 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.controller;
+
+import java.io.IOException;
+
+import javax.servlet.http.HttpServletRequest;
+
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.common.DhisApiVersion;
+import org.hisp.dhis.common.OpenApi;
+import org.hisp.dhis.route.Route;
+import org.hisp.dhis.route.RouteService;
+import org.hisp.dhis.schema.descriptors.RouteSchemaDescriptor;
+import org.hisp.dhis.user.CurrentUser;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.webapi.mvc.annotation.ApiVersion;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.HttpServerErrorException;
+
+/**
+ * @author Morten Olav Hansen
+ */
+@RestController
+@OpenApi.Tags( "integration" )
+@RequiredArgsConstructor
+@RequestMapping( value = RouteSchemaDescriptor.API_ENDPOINT )
+@ApiVersion( { DhisApiVersion.DEFAULT, DhisApiVersion.ALL } )
+public class RouteController
+    extends AbstractCrudController<Route>
+{
+    private final RouteService routeService;
+
+    @RequestMapping( value = "/exec/{id}", method = { RequestMethod.GET, RequestMethod.POST } )
```

Comment:
I don't have a better option but what is `exec` referring to? Perhaps `route` is better.

Reply:
Is `/api/routes/routes` better? it sounds a bit weird

Reply:
Ah sorry. We already have routes from the schema descriptor.

Reply:
What if we add a _new_ controller for the actual "execution" of routes. It could be called something like `ApiGatewayController`, exposed at `/api/gateway`. I noticed that Jira/Atlassian uses that pattern in the their API. Then we define the gateway as a dedicated API, which uses/depends on the RouteService to route requests.

Reply:
We can discuss that later on. I think we should focus on getting it in, we definetely want a lot of input here from the FE side also, I think the important bits is that its clear and understandable what happens behind the endpoint